<a href="https://colab.research.google.com/github/elsyanares/DataAnalyst/blob/main/Market_Basket_Analysis.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Impor packages

In [8]:
# pip install pandas
# pip install numpy
# pip install datetime
import pandas as pd
import numpy as np
import datetime as dt

# Impor data dari CSV ke DataFrame

In [9]:
df = pd.read_csv('/content/Online Retail Data.csv', header=0)
df

order_id product_code                     product_name  quantity  \
0        493410      TEST001          This is a test product.         5   
1       C493411        21539          RETRO SPOTS BUTTER DISH        -1   
2        493412      TEST001          This is a test product.         5   
3        493413        21724  PANDA AND BUNNIES STICKER SHEET         1   
4        493413        84578   ELEPHANT TOY WITH BLUE T-SHIRT         1   
...         ...          ...                              ...       ...   
461768   539991        21618   4 WILDFLOWER BOTANICAL CANDLES         1   
461769   539991        72741            GRAND CHOCOLATECANDLE         4   
461770   539992        21470    FLOWER VINE RAFFIA FOOD COVER         1   
461771   539992        22258          FELT FARM ANIMAL RABBIT         1   
461772   539992        21155            RED RETROSPOT PEG BAG         1   

                 order_date  price  customer_id  
0       2010-01-04 09:24:00   4.50      12346.0  
1       2010-01-04 09:43:00   4.25      14590.0  
2       2010-01-04 09:53:00   4.50      12346.0  
3       2010-01-04 09:54:00   0.85          NaN  
4       2010-01-04 09:54:00   3.75          NaN  
...                     ...    ...          ...  
461768  2010-12-23 16:49:00   1.25          NaN  
461769  2010-12-23 16:49:00   1.45          NaN  
461770  2010-12-23 17:41:00   3.75          NaN  
461771  2010-12-23 17:41:00   1.25          NaN  
461772  2010-12-23 17:41:00   2.10          NaN  

[461773 rows x 7 columns]

In [10]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 461773 entries, 0 to 461772
Data columns (total 7 columns):
 #   Column        Non-Null Count   Dtype  
---  ------        --------------   -----  
 0   order_id      461773 non-null  object 
 1   product_code  461773 non-null  object 
 2   product_name  459055 non-null  object 
 3   quantity      461773 non-null  int64  
 4   order_date    461773 non-null  object 
 5   price         461773 non-null  float64
 6   customer_id   360853 non-null  float64
dtypes: float64(2), int64(1), object(4)
memory usage: 24.7+ MB


# Data cleansing

In [11]:
df_clean = df.copy()
# membuat kolom date
df_clean['date'] = pd.to_datetime(df_clean['order_date']).dt.date.astype('datetime64[ns]')
# menghapus semua baris tanpa customer_id
df_clean = df_clean[~df_clean['customer_id'].isna()]
# menghapus semua baris tanpa product_name
df_clean = df_clean[~df_clean['product_name'].isna()]
# membuat semua product_name berhuruf kecil
df_clean['product_name'] = df_clean['product_name'].str.lower()
# menghapus semua baris dengan product_code atau product_name test
df_clean = df_clean[(~df_clean['product_code'].str.lower().str.contains('test')) |
                    (~df_clean['product_name'].str.contains('test '))]
# membuat kolom order_status dengan nilai 'cancelled' jika order_id diawali dengan huruf 'c' dan 'delivered' jika order_id tanpa awalan huruf 'c'
df_clean['order_status'] = np.where(df_clean['order_id'].str[:1]=='C', 'cancelled', 'delivered')
# mengubah nilai quantity yang negatif menjadi positif karena nilai negatif tersebut hanya menandakan order tersebut cancelled
df_clean['quantity'] = df_clean['quantity'].abs()
# menghapus baris dengan price bernilai negatif
df_clean = df_clean[df_clean['price']>0]
# membuat nilai amount, yaitu perkalian antara quantity dan price
df_clean['amount'] = df_clean['quantity'] * df_clean['price']
# mengganti product_name dari product_code yang memiliki beberapa product_name dengan salah satu product_name-nya yang paling sering muncul
most_freq_product_name = df_clean.groupby(['product_code','product_name'], as_index=False).agg(order_cnt=('order_id','nunique')).sort_values(['product_code','order_cnt'], ascending=[True,False])
most_freq_product_name['rank'] = most_freq_product_name.groupby('product_code')['order_cnt'].rank(method='first', ascending=False)
most_freq_product_name = most_freq_product_name[most_freq_product_name['rank']==1].drop(columns=['order_cnt','rank'])
df_clean = df_clean.merge(most_freq_product_name.rename(columns={'product_name':'most_freq_product_name'}), how='left', on='product_code')
df_clean['product_name'] = df_clean['most_freq_product_name']
df_clean = df_clean.drop(columns='most_freq_product_name')
# mengkonversi customer_id menjadi string
df_clean['customer_id'] = df_clean['customer_id'].astype(str)
# menghapus outlier
from scipy import stats
df_clean = df_clean[(np.abs(stats.zscore(df_clean[['quantity','amount']]))<3).all(axis=1)]
df_clean = df_clean.reset_index(drop=True)
df_clean

order_id product_code                       product_name  quantity  \
0       C493411        21539          red retrospot butter dish         1   
1        493414        21844                  red retrospot mug        36   
2        493414        21533          retro spot large milk jug        12   
3        493414        37508    new england ceramic cake server         2   
4        493414       35001G               hand open shape gold         2   
...         ...          ...                                ...       ...   
358464   539988        84380  set of 3 butterfly cookie cutters         1   
358465   539988       84849D              hot baths soap holder         1   
358466   539988       84849B             fairy soap soap holder         1   
358467   539988        22854        cream sweetheart egg holder         2   
358468   539988       47559B                tea time oven glove         2   

                 order_date  price customer_id       date order_status  amount  
0       2010-01-04 09:43:00   4.25     14590.0 2010-01-04    cancelled    4.25  
1       2010-01-04 10:28:00   2.55     14590.0 2010-01-04    delivered   91.80  
2       2010-01-04 10:28:00   4.25     14590.0 2010-01-04    delivered   51.00  
3       2010-01-04 10:28:00   2.55     14590.0 2010-01-04    delivered    5.10  
4       2010-01-04 10:28:00   4.25     14590.0 2010-01-04    delivered    8.50  
...                     ...    ...         ...        ...          ...     ...  
358464  2010-12-23 16:06:00   1.25     18116.0 2010-12-23    delivered    1.25  
358465  2010-12-23 16:06:00   1.69     18116.0 2010-12-23    delivered    1.69  
358466  2010-12-23 16:06:00   1.69     18116.0 2010-12-23    delivered    1.69  
358467  2010-12-23 16:06:00   4.95     18116.0 2010-12-23    delivered    9.90  
358468  2010-12-23 16:06:00   1.25     18116.0 2010-12-23    delivered    2.50  

[358469 rows x 10 columns]

In [12]:
df_clean.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 358469 entries, 0 to 358468
Data columns (total 10 columns):
 #   Column        Non-Null Count   Dtype         
---  ------        --------------   -----         
 0   order_id      358469 non-null  object        
 1   product_code  358469 non-null  object        
 2   product_name  358469 non-null  object        
 3   quantity      358469 non-null  int64         
 4   order_date    358469 non-null  object        
 5   price         358469 non-null  float64       
 6   customer_id   358469 non-null  object        
 7   date          358469 non-null  datetime64[ns]
 8   order_status  358469 non-null  object        
 9   amount        358469 non-null  float64       
dtypes: datetime64[ns](1), float64(2), int64(1), object(6)
memory usage: 27.3+ MB


# Menyiapkan data basket

## Buat DataFrame basket

In [13]:
basket = pd.pivot_table(df_clean, index='order_id', columns='product_name', values='product_code', aggfunc='nunique', fill_value=0)
basket

product_name  10 colour spaceboy pen  12 ass zinc christmas decorations  \
order_id                                                                  
493414                             0                                  0   
493427                             0                                  0   
493428                             0                                  0   
493432                             0                                  0   
493433                             0                                  0   
...                              ...                                ...   
C539947                            0                                  0   
C539949                            0                                  0   
C539950                            0                                  0   
C539956                            0                                  0   
C539983                            0                                  0   

product_name  12 coloured party balloons  12 daisy pegs in wood box  \
order_id                                                              
493414                                 0                          0   
493427                                 0                          0   
493428                                 0                          0   
493432                                 0                          0   
493433                                 0                          0   
...                                  ...                        ...   
C539947                                0                          0   
C539949                                0                          0   
C539950                                0                          0   
C539956                                0                          0   
C539983                                0                          0   

product_name  12 egg house painted wood  12 ivory rose peg place settings  \
order_id                                                                    
493414                                0                                 0   
493427                                0                                 0   
493428                                0                                 0   
493432                                0                                 0   
493433                                0                                 0   
...                                 ...                               ...   
C539947                               0                                 0   
C539949                               0                                 0   
C539950                               0                                 0   
C539956                               0                                 0   
C539983                               0                                 0   

product_name  12 message cards with envelopes  12 mini toadstool pegs  \
order_id                                                                
493414                                      0                       0   
493427                                      0                       0   
493428                                      0                       0   
493432                                      0                       0   
493433                                      0                       0   
...                                       ...                     ...   
C539947                                     0                       0   
C539949                                     0                       0   
C539950                                     0                       0   
C539956                                     0                       0   
C539983                                     0                       0   

product_name  12 pencil small tube woodland  12 pencils small tube posy  ...  \
order_id                                                            

In [14]:
basket.info()

<class 'pandas.core.frame.DataFrame'>
Index: 19945 entries, 493414 to C539983
Columns: 3862 entries, 10 colour spaceboy pen to zinc willie winkie  candle stick
dtypes: int64(3862)
memory usage: 587.8+ MB


## Encode DataFrame basket dengan nilai True untuk semua nilai di atas 0 dan False untuk semua nilai 0

In [15]:
def encode(x):
    if x==0:
        return False
    if x>0:
        return True

basket_encode = basket.applymap(encode)
basket_encode

/tmp/ipython-input-3017338639.py:7: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  basket_encode = basket.applymap(encode)


product_name  10 colour spaceboy pen  12 ass zinc christmas decorations  \
order_id                                                                  
493414                         False                              False   
493427                         False                              False   
493428                         False                              False   
493432                         False                              False   
493433                         False                              False   
...                              ...                                ...   
C539947                        False                              False   
C539949                        False                              False   
C539950                        False                              False   
C539956                        False                              False   
C539983                        False                              False   

product_name  12 coloured party balloons  12 daisy pegs in wood box  \
order_id                                                              
493414                             False                      False   
493427                             False                      False   
493428                             False                      False   
493432                             False                      False   
493433                             False                      False   
...                                  ...                        ...   
C539947                            False                      False   
C539949                            False                      False   
C539950                            False                      False   
C539956                            False                      False   
C539983                            False                      False   

product_name  12 egg house painted wood  12 ivory rose peg place settings  \
order_id                                                                    
493414                            False                             False   
493427                            False                             False   
493428                            False                             False   
493432                            False                             False   
493433                            False                             False   
...                                 ...                               ...   
C539947                           False                             False   
C539949                           False                             False   
C539950                           False                             False   
C539956                           False                             False   
C539983                           False                             False   

product_name  12 message cards with envelopes  12 mini toadstool pegs  \
order_id                                                                
493414                                  False                   False   
493427                                  False                   False   
493428                                  False                   False   
493432                                  False                   False   
493433                                  False                   False   
...                                       ...                     ...   
C539947                                 False                   False   
C539949                                 False                   False   
C539950                                 False                   False   
C539956                                 False                   False   
C539983                                 False                   False   

product_name  12 pencil small tube woodland  12 pencils small tube posy  ...  \
order_id                                                            

In [16]:
basket_encode.info()

<class 'pandas.core.frame.DataFrame'>
Index: 19945 entries, 493414 to C539983
Columns: 3862 entries, 10 colour spaceboy pen to zinc willie winkie  candle stick
dtypes: bool(3862)
memory usage: 73.6+ MB


## Ambil transaksi dengan banyaknya produk unik lebih dari 1 saja

In [17]:
basket_filter = basket_encode[(basket_encode>0).sum(axis=1)>1]
basket_filter

product_name  10 colour spaceboy pen  12 ass zinc christmas decorations  \
order_id                                                                  
493414                         False                              False   
493427                         False                              False   
493428                         False                              False   
493432                         False                              False   
493433                         False                              False   
...                              ...                                ...   
C539711                        False                              False   
C539719                        False                              False   
C539943                        False                              False   
C539945                        False                              False   
C539950                        False                              False   

product_name  12 coloured party balloons  12 daisy pegs in wood box  \
order_id                                                              
493414                             False                      False   
493427                             False                      False   
493428                             False                      False   
493432                             False                      False   
493433                             False                      False   
...                                  ...                        ...   
C539711                            False                      False   
C539719                            False                      False   
C539943                            False                      False   
C539945                            False                      False   
C539950                            False                      False   

product_name  12 egg house painted wood  12 ivory rose peg place settings  \
order_id                                                                    
493414                            False                             False   
493427                            False                             False   
493428                            False                             False   
493432                            False                             False   
493433                            False                             False   
...                                 ...                               ...   
C539711                           False                             False   
C539719                           False                             False   
C539943                           False                             False   
C539945                           False                             False   
C539950                           False                             False   

product_name  12 message cards with envelopes  12 mini toadstool pegs  \
order_id                                                                
493414                                  False                   False   
493427                                  False                   False   
493428                                  False                   False   
493432                                  False                   False   
493433                                  False                   False   
...                                       ...                     ...   
C539711                                 False                   False   
C539719                                 False                   False   
C539943                                 False                   False   
C539945                                 False                   False   
C539950                                 False                   False   

product_name  12 pencil small tube woodland  12 pencils small tube posy  ...  \
order_id                                                            

In [18]:
basket_filter.info()

<class 'pandas.core.frame.DataFrame'>
Index: 16613 entries, 493414 to C539950
Columns: 3862 entries, 10 colour spaceboy pen to zinc willie winkie  candle stick
dtypes: bool(3862)
memory usage: 61.3+ MB


# Menyiapkan data basket

## Buat DataFrame basket

In [19]:
basket = pd.pivot_table(df_clean, index='order_id', columns='product_name', values='product_code', aggfunc='nunique', fill_value=0)
basket

product_name  10 colour spaceboy pen  12 ass zinc christmas decorations  \
order_id                                                                  
493414                             0                                  0   
493427                             0                                  0   
493428                             0                                  0   
493432                             0                                  0   
493433                             0                                  0   
...                              ...                                ...   
C539947                            0                                  0   
C539949                            0                                  0   
C539950                            0                                  0   
C539956                            0                                  0   
C539983                            0                                  0   

product_name  12 coloured party balloons  12 daisy pegs in wood box  \
order_id                                                              
493414                                 0                          0   
493427                                 0                          0   
493428                                 0                          0   
493432                                 0                          0   
493433                                 0                          0   
...                                  ...                        ...   
C539947                                0                          0   
C539949                                0                          0   
C539950                                0                          0   
C539956                                0                          0   
C539983                                0                          0   

product_name  12 egg house painted wood  12 ivory rose peg place settings  \
order_id                                                                    
493414                                0                                 0   
493427                                0                                 0   
493428                                0                                 0   
493432                                0                                 0   
493433                                0                                 0   
...                                 ...                               ...   
C539947                               0                                 0   
C539949                               0                                 0   
C539950                               0                                 0   
C539956                               0                                 0   
C539983                               0                                 0   

product_name  12 message cards with envelopes  12 mini toadstool pegs  \
order_id                                                                
493414                                      0                       0   
493427                                      0                       0   
493428                                      0                       0   
493432                                      0                       0   
493433                                      0                       0   
...                                       ...                     ...   
C539947                                     0                       0   
C539949                                     0                       0   
C539950                                     0                       0   
C539956                                     0                       0   
C539983                                     0                       0   

product_name  12 pencil small tube woodland  12 pencils small tube posy  ...  \
order_id                                                            

In [20]:
basket.info()

<class 'pandas.core.frame.DataFrame'>
Index: 19945 entries, 493414 to C539983
Columns: 3862 entries, 10 colour spaceboy pen to zinc willie winkie  candle stick
dtypes: int64(3862)
memory usage: 587.8+ MB


## Encode DataFrame basket dengan nilai True untuk semua nilai di atas 0 dan False untuk semua nilai 0

In [21]:
def encode(x):
    if x==0:
        return False
    if x>0:
        return True

basket_encode = basket.applymap(encode)
basket_encode

/tmp/ipython-input-3017338639.py:7: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  basket_encode = basket.applymap(encode)


product_name  10 colour spaceboy pen  12 ass zinc christmas decorations  \
order_id                                                                  
493414                         False                              False   
493427                         False                              False   
493428                         False                              False   
493432                         False                              False   
493433                         False                              False   
...                              ...                                ...   
C539947                        False                              False   
C539949                        False                              False   
C539950                        False                              False   
C539956                        False                              False   
C539983                        False                              False   

product_name  12 coloured party balloons  12 daisy pegs in wood box  \
order_id                                                              
493414                             False                      False   
493427                             False                      False   
493428                             False                      False   
493432                             False                      False   
493433                             False                      False   
...                                  ...                        ...   
C539947                            False                      False   
C539949                            False                      False   
C539950                            False                      False   
C539956                            False                      False   
C539983                            False                      False   

product_name  12 egg house painted wood  12 ivory rose peg place settings  \
order_id                                                                    
493414                            False                             False   
493427                            False                             False   
493428                            False                             False   
493432                            False                             False   
493433                            False                             False   
...                                 ...                               ...   
C539947                           False                             False   
C539949                           False                             False   
C539950                           False                             False   
C539956                           False                             False   
C539983                           False                             False   

product_name  12 message cards with envelopes  12 mini toadstool pegs  \
order_id                                                                
493414                                  False                   False   
493427                                  False                   False   
493428                                  False                   False   
493432                                  False                   False   
493433                                  False                   False   
...                                       ...                     ...   
C539947                                 False                   False   
C539949                                 False                   False   
C539950                                 False                   False   
C539956                                 False                   False   
C539983                                 False                   False   

product_name  12 pencil small tube woodland  12 pencils small tube posy  ...  \
order_id                                                            

In [22]:
basket_encode.info()

<class 'pandas.core.frame.DataFrame'>
Index: 19945 entries, 493414 to C539983
Columns: 3862 entries, 10 colour spaceboy pen to zinc willie winkie  candle stick
dtypes: bool(3862)
memory usage: 73.6+ MB


## Ambil transaksi dengan banyaknya produk unik lebih dari 1 saja

In [23]:
basket_filter = basket_encode[(basket_encode>0).sum(axis=1)>1]
basket_filter

product_name  10 colour spaceboy pen  12 ass zinc christmas decorations  \
order_id                                                                  
493414                         False                              False   
493427                         False                              False   
493428                         False                              False   
493432                         False                              False   
493433                         False                              False   
...                              ...                                ...   
C539711                        False                              False   
C539719                        False                              False   
C539943                        False                              False   
C539945                        False                              False   
C539950                        False                              False   

product_name  12 coloured party balloons  12 daisy pegs in wood box  \
order_id                                                              
493414                             False                      False   
493427                             False                      False   
493428                             False                      False   
493432                             False                      False   
493433                             False                      False   
...                                  ...                        ...   
C539711                            False                      False   
C539719                            False                      False   
C539943                            False                      False   
C539945                            False                      False   
C539950                            False                      False   

product_name  12 egg house painted wood  12 ivory rose peg place settings  \
order_id                                                                    
493414                            False                             False   
493427                            False                             False   
493428                            False                             False   
493432                            False                             False   
493433                            False                             False   
...                                 ...                               ...   
C539711                           False                             False   
C539719                           False                             False   
C539943                           False                             False   
C539945                           False                             False   
C539950                           False                             False   

product_name  12 message cards with envelopes  12 mini toadstool pegs  \
order_id                                                                
493414                                  False                   False   
493427                                  False                   False   
493428                                  False                   False   
493432                                  False                   False   
493433                                  False                   False   
...                                       ...                     ...   
C539711                                 False                   False   
C539719                                 False                   False   
C539943                                 False                   False   
C539945                                 False                   False   
C539950                                 False                   False   

product_name  12 pencil small tube woodland  12 pencils small tube posy  ...  \
order_id                                                            

In [24]:
basket_filter.info()

<class 'pandas.core.frame.DataFrame'>
Index: 16613 entries, 493414 to C539950
Columns: 3862 entries, 10 colour spaceboy pen to zinc willie winkie  candle stick
dtypes: bool(3862)
memory usage: 61.3+ MB


# Mengaplikasikan apriori algorithm



## Buat list frequent itemset (kumpulan produk yang sering dibeli)

1.   List item
2.   List item

In [25]:
# pip install mlxtend
from mlxtend.frequent_patterns import apriori

frequent_itemset = apriori(basket_filter, min_support=.01, use_colnames=True).sort_values('support', ascending=False).reset_index(drop=True)
frequent_itemset['product_cnt'] = frequent_itemset['itemsets'].apply(lambda x: len(x))
frequent_itemset

/usr/local/lib/python3.12/dist-packages/jupyter_client/session.py:203: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
  return datetime.utcnow().replace(tzinfo=utc)


support                                           itemsets  product_cnt
0    0.166075               (white hanging heart t-light holder)            1
1    0.094865                         (regency cakestand 3 tier)            1
2    0.089930                          (jumbo bag red retrospot)            1
3    0.072895                   (strawberry ceramic trinket box)            1
4    0.072473                 (pack of 72 retro spot cake cases)            1
..        ...                                                ...          ...
995  0.010052         (ivory kitchen scales, red kitchen scales)            2
996  0.010052  (white hanging heart t-light holder, metal 4 h...            2
997  0.010052  (red retrospot cake stand, white hanging heart...            2
998  0.010052                              (toadstool money box)            1
999  0.010052  (white hanging heart t-light holder, hanging h...            2

[1000 rows x 3 columns]

/usr/local/lib/python3.12/dist-packages/jupyter_client/session.py:203: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
  return datetime.utcnow().replace(tzinfo=utc)
/usr/local/lib/python3.12/dist-packages/jupyter_client/session.py:203: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
  return datetime.utcnow().replace(tzinfo=utc)
/usr/local/lib/python3.12/dist-packages/jupyter_client/session.py:203: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
  return datetime.utcnow().replace(tzinfo=utc)


## Hitung nilai support, confidence, dan lift dari setiap pasangan produk yang mungkin

In [26]:
from mlxtend.frequent_patterns import association_rules

product_association = association_rules(frequent_itemset, metric='confidence', min_threshold=.7).sort_values(['support','confidence'], ascending=[False,False]).reset_index(drop=True)
product_association

/usr/local/lib/python3.12/dist-packages/jupyter_client/session.py:203: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
  return datetime.utcnow().replace(tzinfo=utc)
/usr/local/lib/python3.12/dist-packages/jupyter_client/session.py:203: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
  return datetime.utcnow().replace(tzinfo=utc)
/usr/local/lib/python3.12/dist-packages/jupyter_client/session.py:203: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
  return datetime.utcnow().replace(tzinfo=utc)
/usr/local/lib/python3.12/dist-packag

antecedents  \
0                  (red hanging heart t-light holder)   
1                    (sweetheart ceramic trinket box)   
2                                 (toilet metal sign)   
3                      (painted metal pears assorted)   
4   (jumbo bag pink with white spots, jumbo bag st...   
..                                                ...   
56  (wood s/3 cabinet ant white finish, wooden pic...   
57  (feltcraft cushion owl, feltcraft cushion rabbit)   
58  (feltcraft cushion owl, feltcraft cushion butt...   
59  (charlotte bag pink with white spots, strawber...   
60                           (key fob , front  door )   

                                          consequents  antecedent support  \
0                (white hanging heart t-light holder)            0.054174   
1                    (strawberry ceramic trinket box)            0.046169   
2                               (bathroom metal sign)            0.024619   
3                     (assorted colour bird ornament)            0.020285   
4                           (jumbo bag red retrospot)            0.020406   
..                                                ...                 ...   
56  (wood 2 drawer cabinet white finish, wooden fr...            0.014326   
57                      (feltcraft cushion butterfly)            0.014146   
58                         (feltcraft cushion rabbit)            0.014386   
59                         (red spotty charlotte bag)            0.013002   
60                                   (key fob , shed)            0.013965   

    consequent support   support  confidence       lift  representativity  \
0             0.166075  0.038885    0.717778   4.322016               1.0   
1             0.072895  0.034912    0.756193  10.373768               1.0   
2             0.036839  0.019623    0.797066  21.636696               1.0   
3             0.071570  0.015410    0.759644  10.613931               1.0   
4             0.089930  0.014928    0.731563   8.134848               1.0   
..                 ...       ...         ...        ...               ...   
56            0.016252  0.010414    0.726891  44.725319               1.0   
57            0.024920  0.010353    0.731915  29.370295               1.0   
58            0.023235  0.010353    0.719665  30.973573               1.0   
59            0.037079  0.010293    0.791667  21.350582               1.0   
60            0.022633  0.010052    0.719828  31.804510               1.0   

    leverage  conviction  zhangs_metric   jaccard  certainty  kulczynski  
0   0.029888    2.954853       0.812651  0.214404   0.661574    0.475960  
1   0.031547    3.802619       0.947340  0.414878   0.737023    0.617568  
2   0.018716    4.746181       0.977856  0.469065   0.789304    0.664873  
3   0.013958    3.862725       0.924539  0.201575   0.741115    0.487475  
4   0.013093    3.390262       0.895342  0.156467   0.705038    0.448780  
..       ...         ...            ...       ...        ...         ...  
56  0.010181    3.602030       0.991851  0.516418   0.722379    0.683816  
57  0.010001    3.637202       0.979812  0.360587   0.725063    0.573687  
58  0.010019    3.484282       0.981839  0.379691   0.712997    0.582631  
59  0.009811    4.622019       0.965719  0.258699   0.783644    0.534632  
60  0.009736    3.488449       0.982275  0.378685   0.713340    0.581988  

[61 rows x 14 columns]

/usr/local/lib/python3.12/dist-packages/jupyter_client/session.py:203: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
  return datetime.utcnow().replace(tzinfo=utc)
/usr/local/lib/python3.12/dist-packages/jupyter_client/session.py:203: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
  return datetime.utcnow().replace(tzinfo=utc)
/usr/local/lib/python3.12/dist-packages/jupyter_client/session.py:203: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
  return datetime.utcnow().replace(tzinfo=utc)


# Kesimpulan


*   Produk Red Hanging Heart T-light Holder dibeli dalam 5,9% dari total transaksi.
* Ketika pembeli sudah membeli Red Hanging Heart T-light Holder, 72% di antaranya juga membeli White Hanging Heart T-light Holder
* Terdapat asosiasi yang kuat antara Red Hanging Heart T-light Holder dan White Hanging Heart T-light Holder. Peluang kedua produk tersebut dibeli bersamaan dalam satu transaksi 4 kali lebih besar dibandingkan dengan peluang keduanya dibeli secara terpisah.
* Berdasarkan insight tersebut, produk White Hanging Heart T-light Holder dapat ditempatkan di dekat/sebelah Red Hanging Heart T-light Holder atau kedua produk dapat dijual dalam bentuk bundel.

